drivers: add support for TeraRanger Evo

msadowski opened the pull request:
Terabee is about to release a new range finder sensor. This commit adds a driver for it.

Any feedback is welcome and highly appreciated.

TSC21 commented on src/drivers/trevo/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+############################################################################
+#
+#   Copyright (c) 2015 PX4 Development Team. All rights reserved.
```
update the year please

TSC21 commented on src/modules/sensors/sensor_params.c:
```diff
@@ -1088,6 +1088,16 @@ PARAM_DEFINE_INT32(SENS_EN_SF0X, 0);
 PARAM_DEFINE_INT32(SENS_EN_MB12XX, 0);
 
 /**
+ * TeraRanger Evo (trevo)
```
Since this is a TeraRanger sensor, I think it would be better to have a parameter `SENS_EN_TR` and add the option to set to 0, if none, 1 if TeraRanger One, 2 if TeraRanger Evo.

TSC21 commented on src/drivers/trevo/trevo.cpp:
```diff
@@ -0,0 +1,938 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2013-2015 PX4 Development Team. All rights reserved.
```
year

TSC21 commented on src/drivers/trevo/trevo.cpp:
```diff
@@ -0,0 +1,938 @@
+/* Configuration Constants */
+#define TREVO_BUS           PX4_I2C_BUS_EXPANSION
+#define TREVO_BASEADDR      0x31 /* 7-bit address */
+#define TREVO_DEVICE_PATH   	"/dev/trevo"
```
fix indentation please

TSC21 commented on src/drivers/trevo/trevo.cpp:
```diff
@@ -0,0 +1,938 @@
+
+/* TREVO Registers addresses */
+
+#define TREVO_MEASURE_REG	0x00		/* Measure range register */
```
fix indentation please

TSC21 commented on src/drivers/trevo/trevo.cpp:
```diff
@@ -0,0 +1,938 @@
+	virtual int 		init();
+
+	virtual ssize_t		read(struct file *filp, char *buffer, size_t buflen);
+	virtual int			ioctl(struct file *filp, int cmd, unsigned long arg);
```
fix indentation please

TSC21 commented on src/drivers/trevo/trevo.cpp:
```diff
@@ -0,0 +1,938 @@
+	ringbuffer::RingBuffer		*_reports;
+	bool				_sensor_ok;
+	uint8_t				_valid;
+	int					_measure_ticks;
```
fix indentation please

TSC21 commented on src/drivers/trevo/trevo.cpp:
```diff
@@ -0,0 +1,938 @@
+	* @param address	The I2C bus address to probe.
+	* @return		True if the device is present.
+	*/
+	int					probe_address(uint8_t address);
```
fix indentation please

TSC21 commented on src/drivers/trevo/trevo.cpp:
```diff
@@ -0,0 +1,938 @@
+	* Perform a poll cycle; collect from the previous measurement
+	* and start a new one.
+	*/
+	void				cycle();
```
fix indentation please

LorenzMeier commented:
Awesome! Good to go if you could address the comments from @TSC21

TSC21 commented:
@msadowski CI is failing on format check so you should address that by fixing style.

ndepal commented:
The only difference between the TR One and TR Evo drivers are the base address and the min/max distances, right?

What do you think about having only one driver that works for both? Either the user chooses which device they want to use it with via parameter (as per @TSC21's suggestion) or we auto-detect which sensor is connected by probing the I2C base addresses.

TSC21 commented:
> auto-detect which sensor is connected by probing the I2C base addresses.

@ndepal that seems an awesome suggestion. @msadowski can you address this instead IOT simplify the driver implementation?

ndepal commented:
When I was testing out the Evo a while back I actually did that. I never got around to testing that the TR One still works with that change and therefore didn't end up making a PR. You can see the implementation here: https://github.com/ndepal/Firmware/commit/85483c2b4d9931ab0878b7c5ae8c981460935e86

One problem might be if there are other devices with the same I2C address. E.g. an Evo and another device with the TR One address. The teraranger driver might then end up thinking that there is a TR One. One way to work around that would be to have the parameter be 0 = off, 1 = autodetect, 2 = TR One, 3 = TR Evo, but I'm not sure if it's worth doing that, it might not be a real issue.

TSC21 commented:
> 0 = off, 1 = autodetect, 2 = TR One, 3 = TR Evo

:+1:

msadowski commented:
@ndepal thanks for your contribution, I can test it tomorrow and give you feedback on it.

We were actually discussing within the team about having one or more drivers. One driver is obviously cleaner but what I'm worrying about are cases when someone would like to use both TeraRanger One and Evo (for example one of them pointing forward). 

@TSC21 If you don't think it's likely to be an issue for anybody I'm more than happy to test @ndepal implementation and if needed fix it and apply it.

TSC21 commented:
I'm more than happy that you do it. Thanks

msadowski commented:
I updated the driver to take the 4 parameters (disabled, automatic, One, Evo), fixed up the copyright dates and tested it with both TeraRanger One and Evo. Everything seems to work now. Please let me know if there is anything you would like us to refactor further.

ndepal commented:
Great! Looks good to me.

TSC21 commented:
@msadowski awesome work! BTW any plans to add a POSIX driver to the TeraRanger products?

msadowski commented:
@TSC21 POSIX driver would be truly neat, I'll ask around to see what others think. I will make sure to record it in our backlog as well.

msadowski commented:
@TSC21: quick follow up: could you please tell me what benefits do you see as a user to have the POSIX driver?

TSC21 commented:
You have different platforms that can there run it - RPi+Navio, Bebop, OcPoC... Or any linux system running PX4 POSIX.
